File: temas/errores/scala/src/main/scala/example/SistemaReservaVuelos.scala
package scala.example

import scala.collection.mutable
import scala.util.{Try, Success, Failure}

object SistemaReservaVuelos {
  
    // Base de datos
    private val vuelos: mutable.Map[String, Vuelo] = mutable.Map()
    private val asientosOriginales: mutable.Map[String, Int] = mutable.Map()
    private val reservas: mutable.Map[String, mutable.Map[String, Reserva]] = mutable.Map()

    // Método para añadir vuelos
    def agregarVuelo(numeroVuelo: String, asientosDisponibles: Int): Either[String, String] = {
        if (numeroVuelo.isEmpty || asientosDisponibles < 0) {
            Left("Detalles del vuelo inválidos.")
        } else {
            vuelos.get(numeroVuelo) match {
                case Some(_) => Left("El vuelo ya existe.")
                case None => 
                    vuelos(numeroVuelo) = Vuelo(numeroVuelo, asientosDisponibles)
                    asientosOriginales(numeroVuelo) = asientosDisponibles
                    Right(s"Vuelo $numeroVuelo agregado exitosamente con $asientosDisponibles asientos.")
            }
        }
    }

    // Método para buscar vuelo
    def buscarVuelo(numeroVuelo: String): Option[Vuelo] = {
        vuelos.get(numeroVuelo)
    }

    // Método para realizar reservas
    def reservarVuelo(numeroVuelo: String, nombrePasajero: String): Either[String, Reserva] = {
        vuelos.get(numeroVuelo) match {
            case Some(vuelo) if vuelo.asientosDisponibles > 0 =>
                val nuevaReserva = Reserva(numeroVuelo, nombrePasajero)
                val reservasVuelo = reservas.getOrElseUpdate(numeroVuelo, mutable.Map())
                reservasVuelo.get(nombrePasajero) match {
                    case Some(_) => Left("Ya existe una reserva para este pasajero.")
                    case None =>
                        vuelos.update(numeroVuelo, vuelo.copy(asientosDisponibles = vuelo.asientosDisponibles - 1))
                        reservasVuelo(nombrePasajero) = nuevaReserva
                        Right(nuevaReserva)
                }
            case Some(_) =>
                Left(s"No hay asientos disponibles en el vuelo $numeroVuelo.")
            case None =>
                Left(s"Vuelo $numeroVuelo no encontrado.")
        }
    }

    // Método para cancelar reserva
    def cancelarReserva(reserva: Reserva): Either[String, String] = {
        if (reserva.numeroVuelo.isEmpty || reserva.nombrePasajero.isEmpty) {
            Left("Detalles de la reserva inválidos.")
        } else {
            vuelos.get(reserva.numeroVuelo) match {
                case Some(vuelo) =>
                    reservas.get(reserva.numeroVuelo).flatMap(_.get(reserva.nombrePasajero)) match {
                        case Some(_) =>
                            vuelos.update(reserva.numeroVuelo, vuelo.copy(asientosDisponibles = vuelo.asientosDisponibles + 1))
                            reservas(reserva.numeroVuelo).remove(reserva.nombrePasajero)
                            Right("Reserva cancelada exitosamente.")
                        case None =>
                            Left("No existe tal reserva.")
                    }
                case None =>
                    Left("Vuelo no encontrado.")
            }
        }
    }

    // Método para eliminar un vuelo
    def eliminarVuelo(numeroVuelo: String): Either[String, String] = {
        vuelos.get(numeroVuelo) match {
            case Some(vuelo) if vuelo.asientosDisponibles == asientosOriginales(numeroVuelo) =>
                vuelos.remove(numeroVuelo)
                asientosOriginales.remove(numeroVuelo)
                reservas.remove(numeroVuelo)
                Right(s"Vuelo $numeroVuelo eliminado exitosamente.")
            case Some(_) =>
                Left("No se puede eliminar el vuelo con reservas activas.")
            case None =>
                Left("Vuelo no encontrado.")
        }
    }
}

File: temas/aspectos/java-02/src/main/java/example/SistemaReservaVuelos.java
package example;

import java.util.HashMap;
import java.util.Map;

public class SistemaReservaVuelos {

    private static final Map<String, Vuelo> vuelos = new HashMap<>();
    private static final Map<String, Integer> asientosOriginales = new HashMap<>();
    private static final Map<String, Map<String, Reserva>> reservas = new HashMap<>();

    public static void reset() {
        vuelos.clear();
        asientosOriginales.clear();
        reservas.clear();
    }

    public static String agregarVuelo(String numeroVuelo, int asientosDisponibles) {
        if (numeroVuelo == null || numeroVuelo.isEmpty() || asientosDisponibles < 0) {
            return "Detalles del vuelo inválidos.";
        } else if (vuelos.containsKey(numeroVuelo)) {
            return "El vuelo ya existe.";
        } else {
            Vuelo vuelo = new Vuelo(numeroVuelo, asientosDisponibles);
            vuelos.put(numeroVuelo, vuelo);
            asientosOriginales.put(numeroVuelo, asientosDisponibles);
            return "Vuelo " + numeroVuelo + " agregado exitosamente con " + asientosDisponibles + " asientos.";
        }
    }

    public static String reservarVuelo(String numeroVuelo, String nombrePasajero) {
        Vuelo vuelo = vuelos.get(numeroVuelo);
        if (vuelo == null) {
            return "Vuelo " + numeroVuelo + " no encontrado.";
        } else if (vuelo.getAsientosDisponibles() > 0) {
            Map<String, Reserva> reservasVuelo = reservas.computeIfAbsent(numeroVuelo, k -> new HashMap<>());
            if (reservasVuelo.containsKey(nombrePasajero)) {
                return "Ya existe una reserva para este pasajero.";
            } else {
                Reserva nuevaReserva = new Reserva(numeroVuelo, nombrePasajero);
                reservasVuelo.put(nombrePasajero, nuevaReserva);
                vuelo.setAsientosDisponibles(vuelo.getAsientosDisponibles() - 1);
                return "Reserva realizada exitosamente.";
            }
        } else {
            return "No hay asientos disponibles en el vuelo " + numeroVuelo + ".";
        }
    }

    public static String cancelarReserva(Reserva reserva) {
        if (reserva.getNumeroVuelo().isEmpty() || reserva.getNombrePasajero().isEmpty()) {
            return "Detalles de la reserva inválidos.";
        }
        Vuelo vuelo = vuelos.get(reserva.getNumeroVuelo());
        if (vuelo == null) {
            return "Vuelo no encontrado.";
        } else {
            Map<String, Reserva> reservasVuelo = reservas.get(reserva.getNumeroVuelo());
            if (reservasVuelo != null && reservasVuelo.remove(reserva.getNombrePasajero()) != null) {
                vuelo.setAsientosDisponibles(vuelo.getAsientosDisponibles() + 1);
                return "Reserva cancelada exitosamente.";
            } else {
                return "No existe tal reserva.";
            }
        }
    }

    public static String eliminarVuelo(String numeroVuelo) {
        Vuelo vuelo = vuelos.get(numeroVuelo);
        if (vuelo == null) {
            return "Vuelo no encontrado.";
        } else if (vuelo.getAsientosDisponibles() == asientosOriginales.getOrDefault(numeroVuelo, 0)) {
            vuelos.remove(numeroVuelo);
            asientosOriginales.remove(numeroVuelo);
            reservas.remove(numeroVuelo);
            return "Vuelo " + numeroVuelo + " eliminado exitosamente.";
        } else {
            return "No se puede eliminar el vuelo con reservas activas.";
        }
    }
}
